Added pole figure plotting capabilities to ExaCA app and make pyebsd optional dependency

gknapp1 opened the pull request:
This adds some features for plotting pole figures using the `pyebsd` package to calculate the Cartesian grid location of the poles. It also creates an `exaca` optional dependency group and moves the import statements to make the `pyebsd` dependency optional.

Added functions:

- The `plot_poles()` function uses the existing `pyebsd.ebsd.plot_PF()` function with minor formatting changes
- The `plot_pole_density()` function uses `pyebsd.ebsd.plot_PF()` to get the location of the poles in Cartesian coordinates and then uses added contouring and plot formatting to generate the pole figure.

gknapp1 commented:
To dos's before re-review:

- [ ] Rebase on main to handle merge conflicts with refactor
- [ ] Correct ExaCA optional dependency error checks

I have a draft of a Gaussian smoothing method for the pole figure plotting, but will leave that for a separate PR.

gknapp1 commented:
> I have a draft of a Gaussian smoothing method for the pole figure plotting, but will leave that for a separate PR.

Since I had it, I went ahead and added the smoothing and some other polishing of the pole figure plot formatting.

MattRolchigo commented on src/myna/application/exaca/export.py:
```diff
@@ -124,3 +126,193 @@ def extract_subregion(
+        from pyebsd.ebsd import plot_PF
+    except ImportError as exc:
+        raise ImportError(
+            'Myna exaca app requires "pip install .[exaca]" optional dependencies!'
```
This error message is kind of confusing - the app requires optional dependencies?

gknapp1 replied:
Yes, the idea is that installing Myna will install the core dependencies and any apps that need additional packages will be optional. For example, if you are installing Myna to use 3DThesis, you will not need whatever dependencies the ExaCA app needs.

Will be more relevant in the case where we have more apps.

MattRolchigo replied:
Oh I think I get it, so if we have apps `A` and `B`, and Myna is using app A which in turn has optional dependencies `a` and `b`, then Myna requires `a` and `b` as well? I think I got confused by the fact that an error message would have "can't find optional dependency" in it, since a missing optional dependency of Myna shouldn't cause an error, but it makes sense if it's referring to an optional dependency of a specific app

gknapp1 replied:
Yep, you've got it right.

MattRolchigo commented on src/myna/application/exaca/export.py:
```diff
@@ -124,3 +126,193 @@ def extract_subregion(
+
+    # Set contour levels
+    if isinstance(levels, int):
+        vmax = np.ceil(np.quantile(hist, 0.99))
```
So this defaults to the "max" of the pole figure being the location on the IPF with the largest estimated multiples of random distribution value? Is this different from what you have in October 30 example pole figure where the MRD max is 18 but no locations seem to exceed 12?

gknapp1 replied:
I think I used the fixed levels option for the MRD max of 18 figure. The fixed levels are more useful if you are comparing across multiple pole figures and want to keep the same scale.

You're right about how this works with the default option. If you say `levels=3` and the max MRD is 20, then you will get `levels = [0.25, 0.5, 0.75, 1, 10, 20]` (levels only refers to levels above 1, levels below 1 are fixed in this mode).

MattRolchigo replied:
I guess it depends on how you're plotting your pole figures and how much smoothing of points you're doing, but having a max MRD of 20 seems really high as a default. Even for directional solidification simulations with no grid-related artifacts, the highest I've seen using the standard MTEX plotting script is around 9.

gknapp1 replied:
By default, the max MRD is calculated from the dataset after it has been smoothed (if smoothing was specified). So if you set `levels = 3` and the max MRD is 5, then the levels will be calculated as `levels = [0.25, 0.5, 0.75, 1, 2.5, 5]`. 

For example plot in this PR, I set `levels = [0.25, 0.5, 0.75, 1, 3, 6, 9, 12, 15, 18]`. Because the function receives a list, it doesn't try to calculate what the maximum is and instead used the exact levels specified.

gknapp1 replied:
I just pushed a change to address the edge case where the maximum count or MRD is <= 1, which previously would have resulted in a crash.

streeve commented:
@MattRolchigo can you review? Could rebase this on main if you want, Gerry

gknapp1 commented:
> @MattRolchigo can you review? Could rebase this on main if you want, Gerry

I went ahead and rebased on main. Would be good if @MattRolchigo could review again. I addressed the comments from his first review.

MattRolchigo commented:
@gknapp1 what's the difference between the functionality of `plot_poles` and `plot_pole_density`? At first I thought one was for inverse pole density figures and the other for pole figures, but it wasn't clear from the code if that's true. Which of these was used to obtain the example pole figure plot?

MattRolchigo commented on src/myna/application/exaca/export.py:
```diff
@@ -124,3 +126,199 @@ def extract_subregion(
+    """Plot the pole figure and return the pole data in Cartesian coordinates
+
+    Args:
+      M: Rotation matrix for sample -> crystal coordinates (passive reference frame)
```
Can you leave a note regarding what the datatype for M is, here and for the other pole function?

gknapp1 replied:
I updated the docstrings accordingly. I also made sure that all other functions in `myna.application.exaca` had docstrings in commit [ba78a67](https://github.com/ORNL-MDF/Myna/pull/51/commits/ba78a675969526a7486c2e9c56cf88c0c264693e)

gknapp1 commented:
> @gknapp1 what's the difference between the functionality of `plot_poles` and `plot_pole_density`? At first I thought one was for inverse pole density figures and the other for pole figures, but it wasn't clear from the code if that's true. Which of these was used to obtain the example pole figure plot?

The pole figure density plot is shown in the merge request description. Will edit to clarify. Both plot pole figures. The `plot_poles` shows the "traditional" version with a point representing each pole.

MattRolchigo commented:
Can `plot_pole_density` allow saving of the figure for consistency with `plot_poles` and `plot_euler_angles`? Also, the output I get from `plot_pole_density` with the defaults doesn't look right - this just looks like a colored version of the pole figure. Do I need to change the defaults? (This was just the top surface from a directional solidification case, so I'd expect a 100 texture)
![PoleDensity](https://github.com/user-attachments/assets/d1294601-96a8-4528-a8e1-bd1d55b2e427)

gknapp1 commented:
> Can `plot_pole_density` allow saving of the figure for consistency with `plot_poles` and `plot_euler_angles`? Also, the output I get from `plot_pole_density` with the defaults doesn't look right - this just looks like a colored version of the pole figure. Do I need to change the default function inputs like `smooth_sigma`? (This was just the top surface from a directional solidification case, so I'd expect a 100 texture) 

The default `bins=256` is a very high resolution of the data, which was set to avoid oversmoothing of the dataset by default. If you have a relatively small dataset compared to `bins^2` then you will see "points" appear since each bin is likely to only have a couple of points. If you want to increase the smoothing, then you can decrease the number of bins.

gknapp1 commented:
> Can `plot_pole_density` allow saving of the figure for consistency with `plot_poles` and `plot_euler_angles`?

It would be best if these functions were consistent. I think it makes more sense to have the behavior be similar to `plot_pole_density` and return the axis/axes with the added plot(s) to allow for modification of the plot formatting by the user, if desired.

MattRolchigo commented:
That makes sense, this dataset only had 40,000 points to plot for the pole figure. I get a smoother representation with fewer bins, and the MRD max starts to converge towards a value (it's a higher value than what MTEX predicts using its default smoothing, but I'm guessing that's due to some subtle nuance of how they bin and plot their data). 

128 bins - MRD max is around 29
64 bins - MRD max is around 16
32 bins - MRD max is around 12
16 bins - MRD max is around 15

What do you think about making a variable default number of bins that is proportional to the input number of grain orientations^2?

gknapp1 commented:
> What do you think about making a variable default number of bins that is proportional to the input number of grain orientations^2?

I've added in a calculated value for `bins` based on the number of data points as the default behavior.


> > Can plot_pole_density allow saving of the figure for consistency with plot_poles and plot_euler_angles?

> It would be best if these functions were consistent. I think it makes more sense to have the behavior be similar to plot_pole_density and return the axis/axes with the added plot(s) to allow for modification of the plot formatting by the user, if desired.

The above is also implemented in the latest commits.

@MattRolchigo please approve if this addresses your comments.
